### ISSUE TITLE
No option to set number of threads for Transformix

### ISSUE THREAD

**Issue by Svdvoort** [InsightSoftwareConsortium/ITKElastix#163]
The `itk.ElastixRegistrationMethod` method has the `SetNumberOfThreads` and `GetNumberOfThreads` functions to set and get the number of threads respectively.

However, `itk.TransformixFilter` does not have such functions. Should the number of threads be set in a different way, or can a `SetNumberOfThreads` function be added?

**Comment by thewtex** [comment]
@N-Dekker

**Comment by N-Dekker** [comment]
@Svdvoort Hi Sebastian, thank you for your suggestion! Do you have a particular use case that would significantly benefit from running itk.TransformixFilter multi-threaded? Would you then expect a large speed-up of the transformation process?

I'm asking, because it may not be entirely trivial to implement this option for TransformixFilter in a thread-safe way. We would have to have a closer look.

**Comment by stefanklein** [comment]
Transformix (the command line version, and I think also the library version) by default already performs multithreading using the default ITK implementation. I think Sebastian's question is just to have a function to set the maximum number of threads, just like possible on the commandline.
Cheers, stefan

**Comment by Svdvoort** [comment]
Indeed, as Stefan points out I would just like to set the maximum number of threads transformix uses. 
I don't expect a speed-up due to using multiple cores with transformix directly.
However, I do have a lot of cores available and would like to run multiple instances of transformix in parallel. 
But because I can't set the (maximum) number of threads at the moment, I'm afraid that a single instance of transformix might use all cores, therefore I would like to be able to set the (maximum) number of threads.

**Comment by N-Dekker** [comment]
@Svdvoort @stefanklein OK, thank you for your clarification. 👍 Please check: https://github.com/SuperElastix/elastix/commit/e5c60f513e5dca8521d38e6ee483dbcab49c2bb8  Of course, once it is merged into SuperElastix/elastix (develop branch), it still needs to be included with the ITKElastix Python wrapping as well.

**Comment by N-Dekker** [comment]
@Svdvoort @thewtex Technically, the TransformixFilter option is there in C++ now, with https://github.com/SuperElastix/elastix/pull/714 (Still to be merged, and wrapped by ITKElastix.) Internally it adds the "-threads" option with the specified number of threads, which triggers calling `itk::MultiThreaderBase::SetGlobalMaximumNumberOfThreads`, just like the transformix executable does. (As was implemented by @stefanklein https://github.com/SuperElastix/elastix/commit/1cf92cb33a8901c08f00ec707e3695a0aee4e815 in 2007).

For the the transformix executable, "-threads" is certainly a useful option. However, when using ITKElastix with the `itk` Python wrapping, doesn't `itk` support calling `itk::MultiThreaderBase::SetGlobalMaximumNumberOfThreads` directly in Python?

**Comment by dzenanz** [comment]
> itk support calling itk::MultiThreaderBase::SetGlobalMaximumNumberOfThreads directly in Python

I guess that would work, but it would affect other filters too (probably not a big deal). I guess it is more intuitive, and easier to find, if the thread limitation exists in the filter.

**Comment by N-Dekker** [comment]
@Svdvoort As a workaround, does it work for you to just call `itk.MultiThreaderBase.SetGlobalMaximumNumberOfThreads` yourself?

We are looking into how to add a function like `transformixFilter.SetNumberOfThreads(numberOfThreads)`, which would set the threads _only_ for that specific `transformixFilter` instance, but it takes a little bit more time.

**Comment by stefanklein** [comment]
Hi Matt,
Strong part of itk-elastix is indeed that we now offer the flexibility of extracting the ITKtransform so that the user can setup their own resampleimagefilter in exactly the way they like.
That having said, I'm not sure if we should turn it around and use it as a reason to not expose functionality via the elastix interface: part of the added value of Elastix is that it wraps the ResampleImageFilter (and GPU accelerated versions of it) and takes care of some book-keeping etc, so that the users don't need to setup the resample image filter themselves.
Cheers, Stefan

**Comment by thewtex** [comment]
Hi @stefanklein ,

Yes, I agree, the flexibility of Elastix helps its adoption. To be clear, I am not suggesting that we do not add the ability to set the number of threads to the Transformix interface. We should continue to increase the flexibility of Elastix so all GPU acceleration versions of resampling, including e.g. Python / CUDA implementations, are available to Elastix users. This is possible by providing an in-memory `itk.Transform` output that can be used with `itk.resample_image_filter`.

**Comment by thewtex** [comment]
Please see the discussion at https://github.com/SuperElastix/elastix/pull/714 on how to provide this functionality on the Transformix interface.

**Comment by stefanklein** [comment]
Thanks for the clarification Matt! Agree. Stefan
